torch/fx/experimental/sym_node.py

```diff
@@ -755,8 +755,75 @@ def _sympy_rshift(a, b):
     return RShift(a, b)
 
 
+def _binary_search_insert_arg(ordered_args, new_arg):
+    """
+    If new_arg is found in ordered_args None is returned, else the new
+    ordered_args with new_arg inserted
+    """
+    low, high = 0, len(ordered_args) - 1
+    from sympy.core.basic import _args_sortkey as sort_key
+
+    # Fast path when new_arg > ordered_args[-1], which is common.
+    if len(ordered_args) and sort_key(ordered_args[-1]) < sort_key(new_arg):
+        return ordered_args + [new_arg]
+
+    while low <= high:
+        mid = (low + high) // 2
+        if ordered_args[mid] == new_arg:
+            return None
+        elif sort_key(ordered_args[mid]) < sort_key(new_arg):
+            low = mid + 1
+        else:
+            high = mid - 1
+
+    ordered_args.insert(low, new_arg)
+    return ordered_args
+
+
+def _optimized_add(
+    lhs, rhs, lhs_is_optimized_summation=False, rhs_is_optimized_summation=False
+):
+    """
+    Custom optimization for Add used to optimize incremental binary summations of certain properties. The idea
+    is when we know the expression is a summation of unique symbols all we need to know is the correct order of symbols,
+    no other optimizations is needed. Figuring out the order can be done in Log(n) comparisons when its known that the input
+    is already sorted. Returns a tuple of (1) a boolean that indicates whether the output is a summation of unique symbols,
+    (2) the result sympy expression.
+    """
+    import sympy
+    from sympy.core.basic import _args_sortkey as sortkey
+
+    def make_optimized(ordered_args):
+        result = sympy.Add(*ordered_args, evaluate=False)
+        return (True, result)
+
+    from torch.utils._sympy.functions import _is_symbols_binary_summation
+
+    lhs_is_optimized_summation = (
+        lhs_is_optimized_summation or _is_symbols_binary_summation(lhs)
+    )
+    rhs_is_optimized_summation = (
+        rhs_is_optimized_summation or _is_symbols_binary_summation(rhs)
+    )
+
+    if lhs_is_optimized_summation:
+        if rhs_is_optimized_summation and sortkey(lhs._args[-1]) < sortkey(
+            rhs._args[0]
+        ):
+            # (a0+a1) + (a2+a3) => (a0+a1+a2+a3)
+            return make_optimized(lhs._args + rhs._args)
```

Comment:
Is it cheap to test the other way too? (You have a cliff here if I accidentally swap the orders of the arguments to successive add here)

Reply:
done